cluster-autoscaler/cloudprovider/gce/mig_target_sizes_provider.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce
+
+import (
+	"fmt"
+	"k8s.io/klog"
+	"sync"
+)
+
+// MigTargetSizesProvider allows obtaining target sizes of MIGs
+type MigTargetSizesProvider interface {
+	// GetMigTargetSize returns targetSize for MIG with given ref
+	GetMigTargetSize(migRef GceRef) (int64, error)
+}
+
+type cachingMigTargetSizesProvider struct {
+	mutex     sync.Mutex
+	cache     *GceCache
+	gceClient AutoscalingGceClient
+	projectId string
+}
+
+// NewCachingMigTargetSizesProvider creates an instance of caching MigTargetSizesProvider
+func NewCachingMigTargetSizesProvider(cache *GceCache, gceClient AutoscalingGceClient, projectId string) MigTargetSizesProvider {
+	return &cachingMigTargetSizesProvider{
+		cache:     cache,
+		gceClient: gceClient,
+		projectId: projectId,
+	}
+}
+
+func (c *cachingMigTargetSizesProvider) GetMigTargetSize(migRef GceRef) (int64, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	targetSize, found := c.cache.GetMigTargetSize(migRef)
+
+	if found {
+		return targetSize, nil
+	}
+
+	newTargetSizes, err := c.fillInMigTargetSizeCache()
+	if err != nil {
+		return 0, err
+	}
+
+	// if we still do not have value here return an error
+	size, found := newTargetSizes[migRef]
+	if !found {
+		return 0, fmt.Errorf("Could not get target size for mig %v", migRef.String())
+	}
+
+	// we are good
+	return size, nil
+}
+
+func (c *cachingMigTargetSizesProvider) fillInMigTargetSizeCache() (map[GceRef]int64, error) {
+	zones := c.listAllZonesForMigs()
+
+	newMigTargetSizeCache := map[GceRef]int64{}
+	for zone := range zones {
+		zoneMigs, err := c.gceClient.FetchAllMigs(zone)
+		if err != nil {
+			klog.Errorf("Error listing migs from zone %v", zone)
+			return nil, err
+		}
+
+		registeredMigRefs := c.getMigRefs()
+
+		for _, zoneMig := range zoneMigs {
+			zoneMigRef := GceRef{
+				c.projectId,
+				zone,
+				zoneMig.Name,
+			}
+
+			if registeredMigRefs[zoneMigRef] {
+				newMigTargetSizeCache[zoneMigRef] = zoneMig.TargetSize
```

Comment:
This feels a bit artificial, you could just SetMigTargetSize here. Tho only difference is if listing fails for one zone we would still update zones we listed previously, which seems reasonable to me.

Reply:
Yeah - yet I still want to return the map from `fillInMigTargetSizeCache` method. So I can check what is the target size of mig without going to cache again (which can be invalidated at this point).
So I would keep it here.

Reply:
Ok, I actually missed the fact you're returning the map. In that case it makes sense.